### PR TITLE
Bytes

### PR DESCRIPTION
**Pull request by HSMF**
Introduce a `bytes` package that mirrors go's `bytes` package. Currently supported functions are `Split` and `Repeat`.

Furthermore, a function `View` in `byteslice` is added to map `[]byte -> seq[byte]`. 

For `Split`, we require the separator to be non-empty. The Go implementation handles this case by returning a string split at every utf8 codepoint boundary: `Split("abc☺", "") -> "a", "b", "c", "☺"`
Do we want to mimic this behavior?

### CODE REVIEW

File: byteslice/byteslice.gobra [hunk 1]
```diff
@@ -33,6 +33,15 @@ func Byte(s []byte, start int, end int, i int) byte {
 	return unfolding acc(Bytes(s, start, end), _) in s[i]
 }
 
+ghost
+requires acc(Bytes(ub, 0, len(ub)), _)
+ensures len(res) == len(ub)
```
Comment (jcp19) [inline byteslice/byteslice.gobra:38]: For consistent styling:
```suggestion
ensures  len(res) == len(ub)
```

File: byteslice/byteslice.gobra [hunk 2]
```diff
@@ -33,6 +33,15 @@ func Byte(s []byte, start int, end int, i int) byte {
...
+ensures forall i int :: { res[i] } 0 <= i && i < len(ub) ==>
+    res[i] == Byte(ub, 0, len(ub), i)
+decreases len(ub)
+opaque
+pure func View(ub []byte) (res seq[byte])
```
Comment (jcp19) [inline byteslice/byteslice.gobra:43]: To make sure that we are not missing anything, could we provide a body to this function here?

File: bytes/bytes_test.gobra [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package bytes
+
+ghost
+decreases
+func testSplit() {
+	abcd := seq[byte]{'a','b','c','d'}
+
+	// this block of assertions is simply to teach gobra how to unfold the definition of `SplitInner`
```
Comment (jcp19) [inline bytes/bytes_test.gobra:8]:
```suggestion
	// these assertions force Gobra to expand the definition of `SplitInner`
```

File: bytes/bytes.gobra [hunk 4]
```diff
@@ -0,0 +1,33 @@
+//+gobra
+
+package bytes
+
```
Comment (jcp19) [inline bytes/bytes.gobra:4]:
```suggestion
```

File: bytes/bytes.gobra [hunk 5]
```diff
@@ -0,0 +1,33 @@
+//+gobra
```
Comment (jcp19) [inline bytes/bytes.gobra:1]: Can you please add a license to these files? The following would be ok:
```
// Any copyright is dedicated to the Public Domain.
// http://creativecommons.org/publicdomain/zero/1.0/
```

File: bytes/bytes.gobra [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ghost
+requires 0 < len(sep)
+decreases len(s)
+pure func SplitInner(s, sep, ac seq[byte]) (res seq[seq[byte]]) {
+	return len(s) == 0 ?
+		( len(ac) == 0 ?
+			seq[seq[byte]]{} :
+			seq[seq[byte]]{ac}) :
+		( sep == s ?
+			seq[seq[byte]]{ ac, seq[byte]{} } :
```
Comment (jcp19) [inline bytes/bytes.gobra:29]: why do we put `seq[byte]{}` in the result here, but not when the accumulator is empty in the base case?

File: bytes/bytes.gobra [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+ghost
+requires 0 < len(sep)
+decreases len(s)
+pure func SplitInner(s, sep, ac seq[byte]) (res seq[seq[byte]]) {
+	return len(s) == 0 ?
+		( len(ac) == 0 ?
+			seq[seq[byte]]{} :
+			seq[seq[byte]]{ac}) :
+		( sep == s ?
```
Comment (jcp19) [inline bytes/bytes.gobra:28]: is this case split necessary? why can't we use the condition used in the next split (`s[:len(sep)] == sep`) here instead?

File: bytes/bytes.gobra (hunk 5 above)
Comment (HSMF) [reply bytes/bytes.gobra:1]: Should I use public domain or the same header as in byteslice.gobra?

File: bytes/bytes.gobra (hunk 6 above)
Comment (HSMF) [reply bytes/bytes.gobra:29]: Good point, that is a bug. `bytes.Split` returns `[[]]` when splitting an empty slice.

**Review by jcp19 — COMMENTED** [review]
after my comments are incorporated, we can merge it. To answer your question, I would keep the precondition that the separator is not empty

File: bytes/bytes.gobra [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+ghost
+requires 0 < len(sep)
+ensures 0 < len(res)
+decreases
+pure func Split(b, sep seq[byte]) (res seq[seq[byte]]) {
+	return SplitInner(b, sep, seq[byte]{})
+}
+
```
[the remark this answers is not in the document]
Comment (HSMF) [reply bytes/bytes.gobra:34]: Do we need to do this manually? The description of the verification step seems imply that every package in the root is checked: "Verify the packages in the root directory"

[the remark this answers is not in the document]
Comment (jcp19) [reply bytes/bytes.gobra:33]: oops, my bad. I had forgotten we set this up in recursive mode

**Comment by HSMF** [comment]
I think this is ready to be merged now

**Pull request merged**